### PR TITLE
[cmake] ADDONS_TO_BUILD exclusion regex and better exact match handling

### LINKED ISSUE #17773: ADDONS_TO_BUILD #17699 causing unwanted libretro addons to build
### Describe the bug
Here is a clear and concise description of what the problem is:

https://github.com/xbmc/xbmc/pull/17699 adds regex arguments for ADDONS_TO_BUILD

This is causing issues in OSMC build script for game.libretro in that all game.libretro.* addons in the repo are being built.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

Only the addons listed should be built. If I use "game.libretro.*" then all matched addons should be built however as there is no * I do not expect it to include all.

## Actual Behavior

 
The build script lists specific game.libretro addons to be built:

```
ADDONS_GAME="game.libretro game.libretro.mame2000 game.libretro.mame2003 game.libretro.scummvm"
```
However this is causing all game.libretro.* addons in the repo to be built which is causing build failures. 

```
-- Pattern game.libretro matches game.libretro.beetle-psx, building addon
```

## Possible Fix

In the PR:

>I think the easier way to resolve this would be to rename game.libretro to something like game.libretro.core. I couldnt come up with anything easy to handle this case, and game.libretro is the only case i could find this happens to across any binary addons. No other active binary addon, has the name that is a subset of other active addons

### To Reproduce

1.  Point build script to list of addons. This contains all addons including ones that that should not be included due to build error/incompatible etc. 
https://github.com/StormTrooper/repo-binary-addons/tree/gmc-19 
2. Run a build

### Debuglog

The debuglog can be found here:

```
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Addon source tarball caching is enabled
-- Building following addons: game.libretro game.libretro.mame2000 game.libretro.mame2003 game.libretro.scummvm

-- Pattern game.libretro matches game.libretro.2048, building addon
-- Pattern game.libretro matches game.libretro.beetle-pce-fast, building addon
-- Pattern game.libretro matches game.libretro.beetle-psx, building addon
-- Pattern game.libretro matches game.libretro.beetle-vb, building addon





```

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [x] Linux
 - [ ] OSX
 - [x] Raspberry-Pi
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name:
Linux raspberrypi4 4.19.75-v7l+ #1270 SMP Tue Sep 24 18:51:41 BST 2019 armv7l GNU/Linux
Running OSMC nightly Matrix build.

 - Kodi version:
https://github.com/xbmc/xbmc/commit/9216e28bc4300eca57dc9671035fc49beb1fa007

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by fuzzard**
## Description

**Commit 1** looks to resolve https://github.com/xbmc/xbmc/issues/17773

examples

```
ADDONS_TO_BUILD='game.libretro peripheral.*'"
```
This will match game.libretro (and only game.libretro), but also match peripheral.joystick and peripheral.xarcade

```
ADDONS_TO_BUILD='game.libretro.* game.libretro peripheral.*'"
```
This will match game.libretro exactly, everything in game.libretro (regex), and both peripheral addons

```
ADDONS_TO_BUILD='game.libretro.mame.* game.libretro peripheral.*'"
```
this will match game.libretro.mame, mame2000, mame2003 (7 mame addons in kodi-game binary addon repo), etc, game.libretro (exact) and then both peripheral addons

```
ADDONS_TO_BUILD='game.libretro.mame game.libretro peripheral.*'"
```
this will match game.libretro.mame (exact), game.libretro (exact) and then both peripheral addons

implements an exclusion match regex to exclude an addon/s from building

example usage
```
ADDONS_TO_BUILD='game.libretro.* -game.libretro.mrboom peripheral.joystick'
```

cmake output
```
-- Exclusion found -game.libretro.2048
-- Addon game.libretro.2048 matches exclusion rule -game.libretro.2048
-- Pattern game.libretro.* matches game.libretro.mrboom, building addon
```
game.libretro.mrboom will not build, but game.libretro and game.libretro.2048 will

example 2:
```
ADDONS_TO_BUILD='game.libretro -game.libretro.mame peripheral.joystick'
```
Cmake output
```
Showing Recent Messages
-- game.libretro.lutro DEPENDENCIES: lutro
-- Addon game.libretro.mame matches exclusion rule game.libretro.mame
-- Addon game.libretro.mame2000 matches exclusion rule game.libretro.mame
-- Addon game.libretro.mame2003 matches exclusion rule game.libretro.mame
-- Addon game.libretro.mame2003_plus matches exclusion rule game.libretro.mame
-- Addon game.libretro.mame2010 matches exclusion rule game.libretro.mame
-- Addon game.libretro.mame2015 matches exclusion rule game.libretro.mame
-- Addon game.libretro.mame2016 matches exclusion rule game.libretro.mame
-- Pattern game.libretro matches game.libretro.melonds, building addon
-- 
```

**Commit 2** Updates docs

## Motivation and Context
Some of the libretro cores can have long build times (eg. game.libretro.mame*).
This allows the ability to build selectively using regex (game.libretro.*) whilst excluding particular addon groups inside that initial regex

## How Has This Been Tested?
ios building kodi-game binary-addon repo and master binary-addon repo

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Review by AlwinEsch — CHANGES_REQUESTED** [review]
About first way on here
-----

Have tested and works nice :smile:

Was on begin not sure if `if(exclusion MATCHES "^[-]${id}")` handle it correct, in case 2 addons have mostly same name e.g. `visualization.milkdrop` and `visualization.milkdrop2`, but has worked on test.

Documentation
-----
The only reason why I marked with "Request changes" is to add somewhere a documentation about, so that a user know this support.
Thought for here:
- https://github.com/xbmc/xbmc/blob/master/cmake/addons/README.md
- And maybe on them here https://github.com/xbmc/xbmc/tree/master/docs
  - not sure how, maybe by reference to `cmake/addons/README.md`

e.g. this on build docs:
```
(3) Build a specific group of add-ons:

make -j$(getconf _NPROCESSORS_ONLN) -C tools/depends/target/binary-addons ADDONS="pvr.*"

For set available styles of ADDONS you can also look by ADDONS_TO_BUILD on here [Kodi add-ons CMake based buildsystem](../cmake/addons/README.md)
```
Or take direct the text from the `cmake/addons/README.md` and copy to this files.

File: cmake/addons/CMakeLists.txt [hunk 1]
```diff
@@ -244,15 +244,40 @@ endif()
 # error either in ADDONS_TO_BUILD or in the directory configuration.
 set(SUPPORTED_ADDON_FOUND FALSE)
 
+set(EXCLUDE_ADDONS "")
+foreach(addon ${ADDONS_TO_BUILD})
+  set(FOUND_EXCLUSION "")
+  string(REGEX MATCH "^[-].*" FOUND_EXCLUSION "${addon}")
+  if(NOT FOUND_EXCLUSION STREQUAL "")
+    # Remove beginning - and then append to excluded addon list
+    STRING(REGEX MATCH "-([^ ]*)"  excluded_addon ${FOUND_EXCLUSION})
```
[the remark this answers is not in the document]
Comment (fuzzard) [reply cmake/addons/CMakeLists.txt:253]: yeah, sorry, had a force push to fix it. Came across the exact issue you first mentioned.

Now i save the excluded_addon list as just eg pvr.dvblink;pvr.dvbviewer without the leading -. Makes the match in the actual addon loop simpler.

[the remark this answers is not in the document]
Comment (fuzzard) [reply cmake/addons/CMakeLists.txt:253]: the match is just the way i see regex matching. Its very loose, and doesnt require regex terms to explicitly force a regex search. ie. -pvr.dvb will match anything with pvr.dvb. Same goes for actual matching ie games.libretro actually matches games.libretro and all games.libretro.*

Honestly i havent tried to fix that situation.

**Comment by AlwinEsch** [comment]
Yes e.g. about example 2, it would be better to have them as `-game.libretro.mame*` and to set with `*` that text is behind.
Without it can have problems that a for creation wanted addon can also be excluded.

**Comment by fuzzard** [comment]
The current matches behaviour hasnt changed with this PR.
At this time, i would see that as out of scope of this PR. I definitely know what you are talking about, but it is also existing behaviour due to the way we use MATCHES

https://github.com/xbmc/xbmc/blob/35641bab4e2d97340269795d0a3a5144aa7538e1/cmake/addons/CMakeLists.txt#L261

The way i see it, you would need to check the supplied string ends in a regex search term (ie. .*), otherwise only do the strict STREQUALS check. Sure it can be done, but it will change current behaviour. Not sure i personally want to get involved in bulk changes, haha.

**Comment by phunkyfish** [comment]
I think just be sure the docs reflect the actual behaviour and that's ok.

**Review by garbear — APPROVED** [review]
Might not have a chance to review the CMake this week but +1 for general concept

**Comment by fuzzard** [comment]
@AlwinEsch when you have time could you have a look at https://github.com/fuzzard/xbmc/commit/f479fb7a0fe6ed57c7535f2fc831ebeb8dae2646

Trying to come up with a way to no regex match a generic term (game.libretro, and vis.milkdrop i your scenario)

It works, but im not sure about it. The branch its in is based on this exclusion branch as well, so it provides

-addon regex exclusion
addon exact match (wont do any further regex matching)
addon not an exact match (then is done using regex matching)

some examples can be seen https://github.com/xbmc/xbmc/issues/17773#issuecomment-620359946

it seems like a hack, as i still have to strip * from anything provided to ensure the filter regex doesnt then match the same as its currently done. Just might be a slight input from things going wild.

The only other way (cleanly) i can come up with would be to change so regex arguments are provides as eg +addonregex, to mirror the -addon for the exclusion state in this rule. Anything without a + or - would then be only considered for exact match. Downside to this, it changes current expected behaviour. At least it would then be explicit in what action a user is wanting however.

**Comment by fuzzard** [comment]
Cleaned it up slightly and made it compatible with cmake 3.5. If/when we do change minimum cmake version to 3.6+, list filters will be handy to change a lot of our looping over to i think.

https://github.com/fuzzard/xbmc/commits/cmake3.5_addonregexmatching

**Comment by fuzzard** [comment]
jenkins build this please

**Comment by fuzzard** [comment]
Have added a commit to resolve https://github.com/xbmc/xbmc/issues/17773

Figure best to try to fix everything in the one PR for this.

**Comment by phunkyfish** [comment]
Are you happy with this on @AlwinEsch ?

**Comment by AlwinEsch** [comment]
Not really, as the most basic no more works `cmake .. -DADDONS_TO_BUILD="screensaver.*"`.
Also seen a bugs inside, give me a bit to make big checks.

**Comment by fuzzard** [comment]
Have you tried it since the new commit? Your use case definitely works with the current PR commits. ie https://github.com/xbmc/xbmc/pull/17755/commits/b2a84a842f1c28db0d714aa7d8a41fd9b0a68641

**Comment by fuzzard** [comment]
@AlwinEsch have updated the commit to your suggested changes. One minor change that just sets ADDONS_TO_BUILD to all if its not set. Save having to check the empty var and string "all" several times

**Comment by fuzzard** [comment]
Jenkins build this please

**Review by AlwinEsch — APPROVED** [review]

**Review by garbear — APPROVED** [review]

**Pull request merged**